lib/fluent/plugin/in_tail.rb

```diff
@@ -630,15 +632,17 @@ def detach_watcher_after_rotate_wait(tw, ino)
             detach_watcher(tw, ino)
           end
         end
-        @tails_rotate_wait[tw] = { ino: ino, timer: timer }
       else
         # when the throttling feature isn't enabled, just wait @rotate_wait
         timer = timer_execute(:in_tail_close_watcher, @rotate_wait, repeat: false) do
           @tails_rotate_wait.delete(tw)
           detach_watcher(tw, ino)
         end
-        @tails_rotate_wait[tw] = { ino: ino, timer: timer }
       end
+
+      # To avoid conflict with timer execution (It deletes the key concurrently)
+      v = @tails_rotate_wait[tw]
+      v[:timer] = timer if v
```

Comment:
> > I am concerned about the possibility of the key being deleted before the key is added. Although I can't confirm such a behavior, I have added the countermeasure just in case.
> > [8b15478](https://github.com/fluent/fluentd/commit/8b1547803f503304ebe8b7beef95ba8a656c701b)
> 
> @ashie Could you briefly check this?

I can't understand what you worry about.
Do you mean that timers created in this method may be executed before reaching here (` v = @tails_rotate_wait[tw]`)?
If so, it's not possible. All timers in in_tail are executed in a same thread.
So created timers are always executed after exiting `detach_watcher_after_rotate_wait` method.

If my above guess for your concern is correct, I don't want this change, I want keep it simple as possible.

Reply:
> Do you mean that timers created in this method may be executed before reaching here (` v = @tails_rotate_wait[tw]`)?
> If so, it's not possible. All timers in in_tail are executed in a same thread. It means that timers created in this method are always executed after exiting this `detach_watcher_after_rotate_wait` method.
> 
> If my above guess for your concern is correct, I don't want this change, I want keep it simple as possible.

Yes!! Thanks so much! I overlooked both `detach_watcher_after_rotate_wait` and the timer event were executed in the same thread for the event loop.

I will remove this fix, and check the following codes as well.

https://github.com/fluent/fluentd/pull/4334/files#diff-456fdeb51bc472beb48891caac0d063e0073655dba7ac2b72e6fdc67dc6ac802R272-R275